### ISSUE TITLE
Oauth2 Grant callback route invalid

### ISSUE THREAD

**Issue by mikkeloscar** [zalando/skipper#1677]
**Describe the bug**

In our Zalando setup the OAuth2 Grant callback looks like this:

```
__oauth2_grant_callback: Path("https://auth.xxxx/.well-known/oauth2-callback")
  -> grantCallback()
  -> <shunt>;
 ```

Which causes the following error logs:

```
[APP]time="2021-01-07T17:04:23Z" level=error msg="* or : in middle of path component https:/auth.xxxx/.well-known/oauth2-callback"
```

One problem is that we are setting a full URL when we should only set a path via the `-oauth2-callback-path`. However, fixing that seems to be not enough as we don't have the hostname which is needed to avoid shadowing other routes in a shared environment like skipper running as ingress controller.

We don't consider the hostname in the route when created here: https://github.com/zalando/skipper/blob/5f5c20274506f0a9bccd2ed8272740acf100b14f/filters/auth/grantprep.go#L15-L33

I guess we need to optionally set a hostname as well to ensure route separating in a shared environment like when skipper is running as ingress controller.

**Comment by szuecs** [comment]
@mikkeloscar no we install only one route as callback. We have no host in this case.

**Comment by AlexanderYastrebov** [comment]
What do you think about having `-oauth2-callback-route` instead of `-oauth2-callback-path`? 
Another idea is to have a flag `-oauth2-callback-disable` which disables default callback route and requires user to specify route manually, e.g. via `-inline-routes`. This would allow flexible configuration of host, path, disable access log, etc.

**Comment by mikkeloscar** [comment]
@szuecs Without host predicate it's possible to shadow backends which also exposes the same path and don't use the skipper oauth grant feature.

I.e. `https://my-app.example.org/.well-known/oauth2-callback` would match even though we may only want `https://auth.example.org/.well-known/oauth2-callback` to match for the oauthGrant feature.

@AlexanderYastrebov I don't have a strong opinion. It seems maybe a bit weird that you enable a feature but have to define the route yourself, on the other hand I can see it's more flexible, but I don't know if this is a need.

**Comment by AlexanderYastrebov** [comment]
My previous idea to config a route instead of path does not help much as `callback-path` is used to construct a redirect url by `oauthGrant` filter as well.

The shadowing problem may be mitigated by using an unique callback path configuration instead of "well-known" `/.well-known/oauth2-callback` that may be used by applications in shared environment.

In general I think the source of the problem is the mismatch between `oauthGrant` and `grantCallback` filters:
*  `oauthGrant` is to be configured by each user/application and in shared environment like in the subject it would be on the route that has Host predicate configured
* `grantCallback` is designed to be automatically injected and would catch requests for all domains

I think it may be rectified such that users must explicitly config `grantCallback` route along with `oauthGrant` route and both would have Host predicate configured (i.e. remove automagic).

**Comment by szuecs** [comment]
@mikkeloscar even with hostname the path could be used by a backend and I don’t think it‘s worth to do it by host.
In the initial PR I also wrote that we might want to have it by host but in general I think it’s ok like it is.

**Comment by mikkeloscar** [comment]
> Can you explain what the current bug is?

My original assumption was that the redirect URL was ALWAYS a fixed on like `https://auth.cluster.com/.well-known/oauth2-callback` (based on how we configured it in our Zalando setup), but I now understand, based on @AlexanderYastrebov comments above and checking the code, that it's intended to be the same host as was used for the `oauthGrant()` i.e. `https://my-app.example.com/.well-known/oauth2-callback`

So with that in mind I understand that it's not a bug (other than we have configured it wrong in our Zalando setup because we specify a full URL instead of just a path).

One last thing that is now clear to me. I was worried that having a global path `/.well-known/oauth2-callback` could shadow the callback URL of some backend which does OAuth2 grant flow but does NOT use the skipper `oauthGrant()` filter. My assumption here was that `/.well-known/oauth2-callback` is a sort of standard, but I fail to find any sources suggesting it is a standard, i.e. it's not listed here: http://www.iana.org/assignments/well-known-uris/well-known-uris.xhtml
So if it's not a standard then it's likely that no backend depends on it and we can roll it out in this way IMO.
